rclcpp/include/rclcpp/parameter_descriptor_wrapper.hpp

```diff
@@ -0,0 +1,102 @@
+#ifndef RCLCPP__PARAMETER_DESCRIPTOR_WRAPPER_HPP_
+#define RCLCPP__PARAMETER_DESCRIPTOR_WRAPPER_HPP_
+
+// Standard library includes
+#include <functional>
+#include <utility>
+#include <memory>
+
+// Additional ROS libraries needed
+#include "node.hpp"
+#include "rcl_interfaces/msg/describe_parameters.hpp"
+#include "rclcpp/node.hpp"
+
+namespace rclcpp
+{
+
+// Implments ParameterDesription class with builder design pattern
+class ParameterDescription
+{
+public:
+    // List of classes the builder manages
+    class FloatingPointDescription;
+
+    // Two constructors in case certain methods don't need to declare_parameters with a Type T in the form of a template
+    ParameterDescription(std::string name, std::uint8_t type, std::string description, std::string additional_constraints, bool read_only. bool dynamic_typing) : m_name{name}, parameter_descriptor.type{type}, m_description{description}, m_additional_constraints{additional_constraints}, read_only{read_only}, dyanmic_typing{m_dyanmic_typing}{};
+
+private:
+    // The main descriptor object we're meant to adjust
+    rcl_interfaces::msg::ParameterDescriptor parameter_descriptor = {};
+
+    // Copies all the information in ParameterDescriptor.msg - https://github.com/ros2/rcl_interfaces/blob/rolling/rcl_interfaces/msg/ParameterDescriptor.msg
+    std::string m_name;
+    std::string m_description;
+
+    // Parameter Constraints
+    std::string m_additional_constraints;
+    bool m_read_only;
+    bool m_dynamic_typing;
+
+};
+
+class ParameterDescription::FloatingPointDescription
+{
+public:
+    // Our Main build methods which will construct the base class
+    ParameterDescription build() const;
+
+    //Builder Method - Mains
+    FloatingPointDescription& SetName(std::string name);
+    FloatingPointDescription& SetType(std::uint8_t type);
+    FloatingPointDescription& SetDescription(std::string description);
+    FloatingPointDescription& SetAdditionalConstraints(std::string constraints);
+    FloatingPointDescription& SetReadOnly(bool read_only);
+    FloatingPointDescription& SetDynamicTyping(bool dynamic_typing);
+
+    // Need the current node in order to begin the configuraiton state forit via the declare_parameter function which setups up the Node
+    template<typename ParameterType>
+    FloatingPointDescription& DeclareParameter<ParameterType>(ParameterType default_value, std::shared_ptr<rclcpp::Node> required_node)
```

Comment:
This shouldn't take a node, but rather the required node interfaces

Reply:
If any implemented class inherits from the `rclcpp::Node` class shouldn't that class be able to fall into the `rclcpp::Node` parameter, or does this engage a bad practice we don't want to follow. The goal is to use the `Node` as a form of `this`, where the original issue suggests [this code on something like line 48](https://git.3dvisionlabs.com/3dvisionlabs/software/hemistereo/apps/stereo/ros2/-/blob/master/src/hemistereo/src/ros2_wrapper.cpp). Where `this` acts as the encasing class which inherits from `rclcpp::Node` seen on [line 20](https://git.3dvisionlabs.com/3dvisionlabs/software/hemistereo/apps/stereo/ros2/-/blob/master/src/hemistereo/includes/ros2_wrapper.hpp) of that file.

Reply:
Not everything that creates a parameter is derived from `rclcpp::Node`; for example `rclcpp_lifecycle::LifecycleNode` doesn't.

The correct approach is to pass the list of node interfaces that you are going to need

Reply:
> Not everything that creates a parameter is derived from `rclcpp::Node`; for example `rclcpp_lifecycle::LifecycleNode` doesn't.
> 
Currently in this pull request I have set up two implementations for the `DeclareParameter` function for both of those sited nodes, after digging around the codebase I saw an implementation [here](https://github.com/alsora/rclcpp/blob/1ce4727ec83e3989122bd5e6a76bad844590dd81/rclcpp/include/rclcpp/parameter_client.hppL#113). Would it be better to follow in the form of template based programming like you had made in that pull request in order to increase viability of the function?

In our case an example would be :
```
template<typename ParameterType, typename NodeT> 

DeclareParameter(ParameterType default_value, std::shared_ptr<NodeT> required_node){
    // Implementation
}

```